File: Python/No10.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if not p: return not s
        first_match = bool(s) and p[0] in {s[0], '.'}

        if len(p) >= 2 and p[1] == '*':
            return (self.isMatch(s, p[2:]) or first_match and self.isMatch(s[1:], p))
        else:
            return first_match and self.isMatch(s[1:], p[1:])


s = Solution()
print(s.isMatch("aa", "a"))
# bool 运算可以将内容进行布尔运算，如果为空内容 ""/0/None/False/[]/{}/()，那么输出为False
# 拓展：如果 p="" 为空，那么 not p == True
# print(bool("ads"))
# print("a" in {"a", '.'})
print((True or False) and False)

File: Java/src/No10.java
class No10 {
    public boolean isMatch(String s, String p) {
        if (p.isEmpty()) return s.isEmpty();
        boolean first_match = (!s.isEmpty() && (p.charAt(0) == s.charAt(0) || p.charAt(0) == '.'));

        if (p.length() >= 2 && p.charAt(1) == '*') {
            return (isMatch(s, p.substring(2)) || (first_match && isMatch(s.substring(1), p)));
        } else {
            return first_match && isMatch(s.substring(1), p.substring(1));
        }
    }

    public static void main(String[] args) {
        System.out.println(new No10().isMatch("mississippi", "mis*is*p*."));
        System.out.println(new No10().isMatch("aab", "c*a*b"));
    }
}

